File: app/backend/controllers/ExamMachineController.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package backend.controllers;

import backend.controllers.base.BaseController;
import backend.models.ExamMachine;
import backend.models.ExamRoom;
import backend.models.Reservation;
import backend.models.Software;
import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.ebean.Ebean;
import io.ebean.Query;
import io.ebean.text.PathProperties;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import org.joda.time.DateTime;
import play.libs.Json;
import play.mvc.Http;
import play.mvc.Result;

public class ExamMachineController extends BaseController {

    @Restrict({ @Group("ADMIN") })
    public Result getExamMachines() {
        List<ExamMachine> machines = Ebean.find(ExamMachine.class).where().eq("archived", false).findList();

        return ok(machines);
    }

    @Restrict({ @Group("ADMIN"), @Group("STUDENT") })
    public Result getExamMachine(Long id) {
        PathProperties pp = PathProperties.parse("(*, softwareInfo(*), room(name, buildingName))");
        Query<ExamMachine> query = Ebean.find(ExamMachine.class);
        pp.apply(query);
        ExamMachine machine = query.where().idEq(id).findOne();
        return ok(machine, pp);
    }

    @Restrict({ @Group("ADMIN") })
    public Result getExamMachineReservationsFromNow(Long id) {
        List<Reservation> reservations = Ebean
            .find(Reservation.class)
            .where()
            .eq("machine.id", id)
            .gt("endAt", DateTime.now()) //.minus(1000 * 60 * 60 * 24))
            .findList();

        return ok(reservations);
    }

    @Restrict({ @Group("ADMIN") })
    public Result updateExamMachine(Long id, Http.Request request) {
        ExamMachine src = formFactory
            .form(ExamMachine.class)
            .bindFromRequest(
                request,
                "id",
                "name",
                "otherIdentifier",
                "accessibilityInfo",
                "accessible",
                "ipAddress",
                "surveillanceCamera",
                "videoRecordings",
                "expanded",
                "statusComment",
                "outOfService"
            )
            .get();
        ExamMachine dest = Ebean.find(ExamMachine.class, id);
        if (dest == null) {
            return notFound();
        }

        if (src.getIpAddress().length() > 0) {
            List<ExamMachine> machines = Ebean.find(ExamMachine.class).findList();
            List<String> ips = machines
                .stream()
                .filter(m -> !m.equals(dest))
                .map(ExamMachine::getIpAddress)
                .collect(Collectors.toList());
            if (ips.contains(src.getIpAddress())) {
                return forbidden("sitnet_error_ip_address_exists_for_room");
            }
        }

        dest.setName(src.getName());
        dest.setOtherIdentifier(src.getOtherIdentifier());
        dest.setAccessibilityInfo(src.getAccessibilityInfo());
        dest.setAccessible(src.isAccessible());
        dest.setIpAddress(src.getIpAddress());
        dest.setSurveillanceCamera(src.getSurveillanceCamera());
        dest.setVideoRecordings(src.getVideoRecordings());
        dest.setExpanded(src.getExpanded());
        dest.setStatusComment(src.getStatusComment());
        dest.setOutOfService(src.getOutOfService());

        dest.update();
        PathProperties pp = PathProperties.parse("(*, softwareInfo(*), room(name, buildingName))");

        return ok(dest, pp);
    }

    @Restrict({ @Group("ADMIN") })
    public Result resetMachineSoftware(Long mid) {
        ExamMachine machine = Ebean.find(ExamMachine.class, mid);
        if (machine == null) {
            return notFound();
        }

        machine.getSoftwareInfo().clear();
        machine.update();
        PathProperties pp = PathProperties.parse("(*, softwareInfo(*), room(name, buildingName))");

        return ok(machine, pp);
    }

    @Restrict({ @Group("ADMIN") })
    public Result updateMachineSoftware(Long mid, Long sid) {
        ExamMachine machine = Ebean.find(ExamMachine.class, mid);
        if (machine == null) {
            return notFound();
        }
        boolean isTurnedOn = false;
        Software software = Ebean.find(Software.class, sid);
        if (machine.getSoftwareInfo().contains(software)) {
            machine.getSoftwareInfo().remove(software);
        } else {
            machine.getSoftwareInfo().add(software);
            isTurnedOn = true;
        }
        machine.update();
        ObjectNode part = Json.newObject();
        part.put("turnedOn", isTurnedOn);

        return ok(Json.toJson(part));
    }

    @Restrict({ @Group("ADMIN") })
    public Result insertExamMachine(Long id, Http.Request request) {
        ExamRoom room = Ebean.find(ExamRoom.class, id);
        if (room == null) {
            return notFound();
        }
        ExamMachine machine = bindForm(ExamMachine.class, request);
        room.getExamMachines().add(machine);
        room.save();

        machine.save();

        return ok(machine);
    }

    @Restrict(@Group({ "ADMIN" }))
    public Result removeExamMachine(Long id) {
        // SIT-690
        return forbidden();
    }

    @Restrict({ @Group("TEACHER"), @Group("ADMIN"), @Group("STUDENT") })
    public Result getSoftwares() {
        List<Software> softwares = Ebean
            .find(Software.class)
            .where()
            .or()
            .isNull("status")
            .eq("status", "ACTIVE")
            .endOr()
            .findList();

        return ok(softwares);
    }

    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result getSoftware(Long id) {
        Software software = Ebean.find(Software.class, id);

        return ok(software);
    }

    private Optional<Result> checkSoftwareName(String name) {
        List<Software> sw = Ebean.find(Software.class).where().ieq("name", name).findList();
        return sw.isEmpty() ? Optional.empty() : Optional.of(badRequest("Software with that name already exists"));
    }

    @Restrict(@Group({ "ADMIN" }))
    public Result addSoftware(String name, Http.Request request) {
        Software software = bindForm(Software.class, request);
        return checkSoftwareName(name)
            .orElseGet(
                () -> {
                    software.setName(name);
                    software.save();
                    return ok(software);
                }
            );
    }

    @Restrict(@Group({ "ADMIN" }))
    public Result updateSoftware(Long id, String name) {
        Software software = Ebean.find(Software.class, id);
        if (software == null) {
            return notFound();
        }
        return checkSoftwareName(name)
            .orElseGet(
                () -> {
                    software.setName(name);
                    software.update();
                    return ok(software);
                }
            );
    }

    @Restrict(@Group({ "ADMIN" }))
    public Result removeSoftware(Long id) {
        Software software = Ebean.find(Software.class, id);
        if (software == null) {
            return notFound();
        }
        software.delete();
        return ok();
    }
}

File: app/features/facility/controllers/ExamMachineController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package features.facility.controllers

import database.EbeanJsonExtensions
import features.facility.services.{ExamMachineError, ExamMachineService}
import models.user.Role
import play.api.libs.json.{JsValue, Json}
import play.api.mvc.*
import security.Auth.{AuthenticatedAction, authorized}
import security.BlockingIOExecutionContext
import system.AuditedAction

import javax.inject.Inject

class ExamMachineController @Inject() (
    private val examMachineService: ExamMachineService,
    val authenticated: AuthenticatedAction,
    val audited: AuditedAction,
    val controllerComponents: ControllerComponents,
    implicit val ec: BlockingIOExecutionContext
) extends BaseController
    with EbeanJsonExtensions:

  private def toResult(error: ExamMachineError): Result =
    error match
      case ExamMachineError.MachineNotFound  => NotFound(ExamMachineError.MachineNotFound.message)
      case ExamMachineError.RoomNotFound     => NotFound(ExamMachineError.RoomNotFound.message)
      case ExamMachineError.SoftwareNotFound => NotFound(ExamMachineError.SoftwareNotFound.message)
      case ExamMachineError.IpAddressConflict(msg)    => Forbidden(msg)
      case ExamMachineError.SoftwareNameConflict(msg) => BadRequest(msg)

  def getExamMachines: Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.SUPPORT))) { _ =>
      val machines = examMachineService.getExamMachines
      Ok(machines.asJson)
    }

  def getExamMachine(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN, Role.Name.STUDENT))) { _ =>
      examMachineService.getExamMachine(id) match
        case Left(error)          => toResult(error)
        case Right((machine, pp)) => Ok(machine.asJson(pp))
    }

  def getExamMachineReservationsFromNow(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      val reservations = examMachineService.getExamMachineReservationsFromNow(id)
      Ok(reservations.asJson)
    }

  def updateExamMachine(id: Long): Action[JsValue] =
    audited.andThen(authenticated)(parse.json).andThen(authorized(Seq(Role.Name.ADMIN))) {
      request =>
        examMachineService.updateExamMachine(id, request.body) match
          case Left(error)          => toResult(error)
          case Right((machine, pp)) => Ok(machine.asJson(pp))
    }

  def resetMachineSoftware(mid: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      examMachineService.resetMachineSoftware(mid) match
        case Left(error)          => toResult(error)
        case Right((machine, pp)) => Ok(machine.asJson(pp))
    }

  def updateMachineSoftware(mid: Long, sid: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      examMachineService.updateMachineSoftware(mid, sid) match
        case Left(error)       => toResult(error)
        case Right(isTurnedOn) => Ok(Json.obj("turnedOn" -> isTurnedOn))
    }

  def insertExamMachine(id: Long): Action[AnyContent] =
    audited.andThen(authenticated).andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      examMachineService.insertExamMachine(id) match
        case Left(error)    => toResult(error)
        case Right(machine) => Ok(machine.asJson)
    }

  def removeExamMachine(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      examMachineService.removeExamMachine(id) match
        case Left(error) => toResult(error)
        case Right(_)    => Ok
    }

  def listSoftware: Action[AnyContent] =
    authenticated { _ =>
      val software = examMachineService.listSoftware
      Ok(software.asJson)
    }

  def getSoftware(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT))) {
      _ =>
        examMachineService.getSoftware(id) match
          case Left(error)     => toResult(error)
          case Right(software) => Ok(software.asJson)
    }

  def addSoftware(name: String): Action[AnyContent] =
    audited.andThen(authenticated).andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      examMachineService.addSoftware(name) match
        case Left(error)     => toResult(error)
        case Right(software) => Ok(software.asJson)
    }

  def updateSoftware(id: Long, name: String): Action[AnyContent] =
    audited.andThen(authenticated).andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      examMachineService.updateSoftware(id, name) match
        case Left(error)     => toResult(error)
        case Right(software) => Ok(software.asJson)
    }

  def removeSoftware(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.ADMIN))) { _ =>
      examMachineService.removeSoftware(id) match
        case Left(error) => toResult(error)
        case Right(_)    => Ok
    }
